Mathlib/CategoryTheory/Limits/MonoCoprod.lean

```diff
@@ -107,6 +110,106 @@ instance monoCoprodType : MonoCoprod (Type u) :=
+  BinaryCofan.mk (Cofan.IsColimit.desc hc₁ (fun i₁ => c.inj (Sum.inl i₁)))
+    (Cofan.IsColimit.desc hc₂ (fun i₂ => c.inj (Sum.inr i₂)))
+
+/-- The binary cofan `binaryCofanSum c c₁ c₂ hc₁ hc₂` is colimit. -/
+def isColimitBinaryCofanSum : IsColimit (binaryCofanSum c c₁ c₂ hc₁ hc₂) :=
+  BinaryCofan.IsColimit.mk _ (fun f₁ f₂ => Cofan.IsColimit.desc hc (fun i => match i with
+      | Sum.inl i₁ => c₁.inj i₁ ≫ f₁
+      | Sum.inr i₂ => c₂.inj i₂ ≫ f₂))
+    (fun f₁ f₂ => Cofan.IsColimit.hom_ext hc₁ _ _ (by simp))
+    (fun f₁ f₂ => Cofan.IsColimit.hom_ext hc₂ _ _ (by simp))
+    (fun f₁ f₂ m hm₁ hm₂ => by
+      apply Cofan.IsColimit.hom_ext hc
+      rintro (i₁|i₂) <;> aesop_cat)
+
+lemma mono_binaryCofanSum_inl [MonoCoprod C] :
+    Mono (binaryCofanSum c c₁ c₂ hc₁ hc₂).inl :=
+  MonoCoprod.binaryCofan_inl _ (isColimitBinaryCofanSum c c₁ c₂ hc hc₁ hc₂)
+
+lemma mono_binaryCofanSum_inr [MonoCoprod C] :
+    Mono (binaryCofanSum c c₁ c₂ hc₁ hc₂).inr :=
+  MonoCoprod.binaryCofan_inr _ (isColimitBinaryCofanSum c c₁ c₂ hc hc₁ hc₂)
+
+lemma mono_binaryCofanSum_inl' [MonoCoprod C] (inl : c₁.pt ⟶ c.pt)
+    (hinl : ∀ (i₁ : I₁), c₁.inj i₁ ≫ inl = c.inj (Sum.inl i₁)) :
+    Mono inl := by
+  suffices inl = (binaryCofanSum c c₁ c₂ hc₁ hc₂).inl by
+    rw [this]
+    exact MonoCoprod.binaryCofan_inl _ (isColimitBinaryCofanSum c c₁ c₂ hc hc₁ hc₂)
+  exact Cofan.IsColimit.hom_ext hc₁ _ _ (by simpa using hinl)
+
+lemma mono_binaryCofanSum_inr' [MonoCoprod C] (inr : c₂.pt ⟶ c.pt)
+    (hinr : ∀ (i₂ : I₂), c₂.inj i₂ ≫ inr = c.inj (Sum.inr i₂)) :
+    Mono inr := by
+  suffices inr = (binaryCofanSum c c₁ c₂ hc₁ hc₂).inr by
+    rw [this]
+    exact MonoCoprod.binaryCofan_inr _ (isColimitBinaryCofanSum c c₁ c₂ hc hc₁ hc₂)
+  exact Cofan.IsColimit.hom_ext hc₂ _ _ (by simpa using hinr)
+
+end
+
+section
+
+variable [MonoCoprod C] {I J : Type*} (X : I → C) (ι : J → I) (hι : Function.Injective ι)
+
+section
+
+variable (c : Cofan X) (c₁ : Cofan (X ∘ ι)) (hc : IsColimit c) (hc₁ : IsColimit c₁)
+
+lemma mono_of_injective_aux (c₂ : Cofan (fun (k : ((Set.range ι)ᶜ : Set I)) => X k.1))
+    (hc₂ : IsColimit c₂) : Mono (Cofan.IsColimit.desc hc₁ (fun i => c.inj (ι i))) := by
+  classical
+  let e := ((Equiv.ofInjective ι hι).sumCongr (Equiv.refl _)).trans (Equiv.Set.sumCompl _)
+  refine mono_binaryCofanSum_inl' (Cofan.mk c.pt (fun i' => c.inj (e i'))) _ _ ?_
+    hc₁ hc₂ _ (by simp)
+  exact IsColimit.ofIsoColimit ((IsColimit.ofCoconeEquiv (Cocones.equivalenceOfReindexing
+    (Discrete.equivalence e) (Iso.refl _))).symm hc) (Cocones.ext (Iso.refl _))
+
+lemma mono_of_injective [HasCoproduct (fun (k : ((Set.range ι)ᶜ : Set I)) => X k.1)] :
+    Mono (Cofan.IsColimit.desc hc₁ (fun i => c.inj (ι i))) :=
+  mono_of_injective_aux X ι hι c c₁ hc hc₁ _ (colimit.isColimit _)
+
+end
+
+lemma mono_of_injective' [HasCoproduct (X ∘ ι)] [HasCoproduct X]
+    [HasCoproduct (fun (k : ((Set.range ι)ᶜ : Set I)) => X k.1)] :
+    Mono (Sigma.desc (f := X ∘ ι) (fun i => Sigma.ι X (ι i))) :=
```

Comment:
I wonder if we should prefer the spelling
```suggestion
    Mono (Sigma.map' ι (fun j => 𝟙 (X ∘ ι) j)) :=
```
here. The two maps are not defeq as the `map'` spelling introduces a composition with an identity that's not there in your version, but `map'` is supposed to be the general "map sigma into other sigma", so it would be nice to have your result for `map'`. Maybe we want both?

Reply:
Thanks for the suggestion! I have added the `map'` version.